Automattic/woocommerce-payments
# Fix to an issue with attaching exchange rates for zero-decimal currencies

Fixes #2930

#### Changes proposed in this Pull Request





- Adds a fix to ensure the exchange rate for zero-decimal to non-zero-decimal currencies is stored correctly. The exchange rate provided by Stripe needs to be multipled/divided by 100 in this case, because Stripe expects non-zero decimal amounts to be provided in cents (e.g. 1000 USD = $10.00), but zero-decimal currencies are provided as-is (e.g. 1000 JPY = 1000 yen). This leads to the exchange rate being off by 2dp in these cases.
- Also added a check to make sure the Stripe account currency matches the store default currency. If there is a mis-match here, the exchange rate stored in Stripe will not be usable for analytics because it is converting between presentment currency and Stripe account currency, where we are expecting it to be between presentment currency and store default currency. In these cases, we can just fallback to using the `_wcpay_multi_currency_order_exchange_rate meta key for analytics.
  
- 



#### Testing instructions





* Go to `Settings -> Multi Currency` and add Japanese Yen to your enabled currencies. If the exchange rate isn't fetched manually from the server (this could happen if you are pointing at your local WCPay Server) then go into the settings for Yen and set a manual rate (check what the USD -> Yen conversion rate is using a currency conversion website).
* Also make sure you have a non-zero decimal currency like GBP or EUR set up in Multi-Currency.
* Go to your store, change the currency to JPY using the multi-currency dropdown, and then purchase a product. Make a note of the value of the order in JPY and the original value in USD.
* Change the currency to EUR/GBP using the multi-currency dropdown and make another order. Note the value of the order in EUR and the original value in USD.
* Go to `localhost:8083` and check the `wp_postmeta` table. You should see a meta key `_wcpay_multi_currency_stripe_exchange_rate` for the two orders.
* For the JPY order, multiply the exchange rate by the order value and make sure it gives you the original value of the order in USD (note: this may not be exact, because of fluctuations and differences between the Stripe exchange rate and the multi-currency exchange rate, but it should be close).
* Do the same with the EUR order, and verify it gives you a value which is close to the USD order value.
* Go to `Analytics -> Orders` and verify that you can see the two orders and the totals are displayed correctly.

-------------------


**Post merge**
